src/js/background.ts

```diff
@@ -92,13 +115,12 @@ function handleMessages(
       ).then(validationResult => {
         if (validationResult.valid) {
           const manifestMap = getManifestMapForOrigin(message.origin);
-          let manifest = manifestMap.get(message.version);
-          if (!manifest) {
-            manifest = {
-              leaves: [],
-              root: message.rootHash,
-              start: Date.now(),
-            };
+          const manifest = manifestMap.get(message.version) ?? {
+            leaves: [],
+            root: message.rootHash,
+            start: Date.now(),
+          };
+          if (!manifestMap.has(message.version)) {
```

Comment:
Feels to me like the previous version of the code was safer?

What happens when manifestMap is null on line 123? Does the browser barf or does it handle it gracefully?

Reply:
I believe these two are identical in logic. In the first one:
1. We read a value from the manifest, which might be undefined.
2. If it is undefined we **assign a new default value to `manifest`** and set that value back into the map.

In the new version
1. We create a `const manifest` that is equal to the value in the map **or a default value**.
2. If no value exists for this origin in the map, we assign manifest to it.

So we're changing where the defaulting happens and how we're checking whether the key exists, but all the steps are ultimately the same. And in both cases we're guaranteed to have a valid manifest.

As for **why**, this is a way of getting around a weird limitation in TypeScript. Under normal circumstances the original code works fine, and TS will understand that in code after the `if` block, `manifest` is guaranteed to be defined.

Unfortunately it gets weird with the `forEach` loop below. For some reason TS assumes that something could modify `manifest` because it is a `let` variable. It could theoretically know that the callback to the `forEach` doesn't modify the `manifest` variable, but I guess it's probably some performance optimization for the compiler?

Regardless, changing this from `let` to `const` fixes the issue.

Reply:
`manifestMap` should be non null, not sure I follow

Reply:
`manifestMap` is fine, it's `manifest` that TS has issues with. On line 127/128 (of the new code) we try to access `manifest.leaves`, but due to the issue I mentioned, TS think `manifest` might not exist at all.

Here's a minimal reproduction of the issue. Not how with `let` even though we're accessing `foo` inside a block where we literally just told it `foo !== undefined`, it still thinks it could be undefined inside the callback to `forEach`. Whereas with `const` it correctly narrows.

```js
const map: string[] = [];

let foo: string | undefined = 'foo';
if (foo !== undefined) {
  map.forEach(item => foo); // foo: string | undefined
}

const bar: string | undefined = 'bar';
if (bar !== undefined) {
  map.forEach(item => bar); // bar: string
}
```